api/genericresource/resource_management.go

```diff
@@ -0,0 +1,62 @@
+package genericresource
+
+import (
+	"github.com/docker/swarmkit/api"
+)
+
+// Claim assigns GenericResources to the task by taking them from the node's GenericResource
+// and storing them in the taken store
+func Claim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
+	t, err := getGenericResourceDiscrete(taskRes)
+	if err != nil {
+		return err
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.GenericResource_Discrete:
+		if nr.Discrete < t.Discrete {
+			return NewError("cannot provide enough of scalar resource requested %d, available %d",
+				t.Discrete, nr.Discrete)
+		}
+		nr.Discrete -= t.Discrete
+		taken.Resource = NewDiscrete(t.Discrete)
+	case *api.GenericResource_Set:
+		i := uint64(0)
+		res := NewSet()
+
+		for k := range nr.Set.Val {
+			if i >= t.Discrete {
```

Comment:
It *does* differentiate them as you can't replicate exclusive access
If it were expressed as an integer then you wouldn't have any way to tell
which disk was given to you.

The example I am showing in the linked demo is the same:
- node 1 has {GPU-0} and label maxwell
- node 2 has {GPU-0} and label pascal
- node 2 has {GPU-0, GPU-1} and label pascal

When I am creating tasks, requesting a GPU, the scheduler:
- picks a GPU from the set
- Gives the task the name of the item it selected through en environment variable

If this were expressed through an integer then I wouldn't be able to tell which GPU was assigned to me.